File: 2021/06/11/main.go
package main

func stoneGameVII(stones []int) int {
	d := len(stones)
	dp := make([][]int, d)
	for i := range dp {
		dp[i] = make([]int, d)
	}
	for i := d - 1; i > -1; i-- {
		sm := stones[i]
		for j := i + 1; j < d; j++ {
			sm += stones[j]
			dp[i][j] = max(sm-stones[i]-dp[i+1][j], sm-stones[j]-dp[i][j-1])
		}
	}
	return dp[0][d-1]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 2021/06/11/main.py
class Solution:
    def stoneGameVII(self, stones: List[int]) -> int:
        d = len(stones)
        dp = [[0] * d for _ in range(d)]
        for i in range(d-1, -1, -1):
            sm = stones[i]
            for j in range(i+1, d):
                sm += stones[j]
                dp[i][j] = max(sm-stones[i]-dp[i+1][j], sm-stones[j]-dp[i][j-1])
        return dp[0][-1]
